File: clud/plugin/token_resolver.py
"""Resolve the Claude Code OAuth access token from where Claude Code stores it.

WHY local token (not a claude.ai cookie): clud runs inside the Claude Code
environment; reusing the same token Claude Code's own /usage uses needs zero
manual setup. See the spec's compliance note — this is an undocumented,
best-effort path; every failure degrades to "usage unavailable".

SECURITY: the raw token value is never logged and never returned on a failure
path. TokenResult carries only whether a (live) token was found, its source,
and a machine reason — safe to log/diagnose.
"""

from __future__ import annotations

import json
import subprocess
import time
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

KEYCHAIN_SERVICE = "Claude Code-credentials"
CREDENTIALS_PATH = Path("~/.claude/.credentials.json").expanduser()
_MS_THRESHOLD = 1_000_000_000_000  # values below this look like epoch seconds
# Lower bound for the seconds-coercion: a *plausible* epoch-seconds timestamp is
# >= 1e9 s (year 2001). Only values in [1e9, 1e12) are coerced s->ms; anything
# smaller (e.g. a tiny test/sentinel value) is taken as already-ms so a clearly
# past timestamp still reads as expired instead of being inflated into the future.
_SECONDS_FLOOR = 1_000_000_000


@dataclass(frozen=True)
class TokenResult:
    token: str | None
    source: str  # "env" | "file" | "keychain" | "none"
    reason: str | None  # None on success; "no_token" | "expired"


def _read_keychain(service: str) -> str | None:
    """Read a generic-password value from the macOS login Keychain, or None.

    First access of an item another app created may prompt the user for
    authorization (one-time "Always Allow"); a denial / missing item / any
    error just yields None → usage degrades to unavailable, never crashes.
    """
    try:
        proc = subprocess.run(
            ["security", "find-generic-password", "-s", service, "-w"],
            capture_output=True,
            text=True,
            timeout=5,
        )
    except (OSError, subprocess.SubprocessError):
        return None
    if proc.returncode != 0:
        return None
    return proc.stdout.strip() or None


def _parse_oauth_blob(text: str) -> tuple[str | None, int | None]:
    """Extract (accessToken, expiresAt_ms) from a credentials JSON blob.

    Shape: {"claudeAiOauth": {"accessToken": "...", "expiresAt": <epoch ms>}}.
    Returns (None, None) for any malformed input.
    """
    try:
        doc = json.loads(text)
    except (json.JSONDecodeError, ValueError):
        return None, None
    oauth = doc.get("claudeAiOauth") if isinstance(doc, dict) else None
    if not isinstance(oauth, dict):
        return None, None
    token = oauth.get("accessToken")
    expires = oauth.get("expiresAt")
    token = token if isinstance(token, str) and token else None
    expires_ms: int | None = None
    if isinstance(expires, (int, float)) and not isinstance(expires, bool):
        expires_ms = int(expires)
        # Guard a seconds-valued field so we never false-flag a live token, but
        # only when the value is a plausible epoch-seconds timestamp ([1e9, 1e12)).
        if _SECONDS_FLOOR <= expires_ms < _MS_THRESHOLD:
            expires_ms *= 1000
    return token, expires_ms


def resolve_token(
    *,
    env: dict[str, str] | None = None,
    credentials_path: Path | None = None,
    keychain_reader: Callable[[str], str | None] | None = None,
    now_ms: int | None = None,
) -> TokenResult:
    """Resolve a live OAuth token, or a TokenResult explaining why not.

    Re-call this every fetch: a token Claude Code rotates on disk is picked
    up automatically. The seams (env / credentials_path / keychain_reader /
    now_ms) are injected so the resolver is fully unit-testable offline.
    """
    import os

    env = env if env is not None else dict(os.environ)
    credentials_path = credentials_path or CREDENTIALS_PATH
    keychain_reader = keychain_reader or _read_keychain
    now_ms = now_ms if now_ms is not None else int(time.time() * 1000)

    env_tok = env.get("CLAUDE_CODE_OAUTH_TOKEN")
    if env_tok:
        return TokenResult(token=env_tok, source="env", reason=None)

    token: str | None = None
    expires_ms: int | None = None
    source = "none"
    try:
        token, expires_ms = _parse_oauth_blob(credentials_path.read_text())
        if token is not None:
            source = "file"
    except (FileNotFoundError, OSError):
        pass

    if token is None:
        blob = keychain_reader(KEYCHAIN_SERVICE)
        if blob:
            token, expires_ms = _parse_oauth_blob(blob)
            if token is not None:
                source = "keychain"

    if token is None:
        return TokenResult(token=None, source="none", reason="no_token")
    if expires_ms is not None and expires_ms <= now_ms:
        return TokenResult(token=None, source=source, reason="expired")
    return TokenResult(token=token, source=source, reason=None)

File: clusage-menubar/Sources/ClusageCore/TokenResolver.swift
/// TokenResolver.swift — resolve the Claude Code OAuth token from where Claude Code stores it.
///
/// WHY local token (not a claude.ai cookie): Clusage reuses the same token Claude
/// Code's own /usage uses; zero manual setup for the user. See README compliance note —
/// undocumented endpoint, degrades gracefully to unavailable on every failure.
///
/// SECURITY: the raw token value is never logged and never returned on a failure path.
/// TokenResult carries only whether a live token was found, its source, and a machine
/// reason — safe to log/diagnose.
///
/// Port of clud/plugin/token_resolver.py with one critical addition: the multi-item
/// Keychain selection fix (two items share "Claude Code-credentials" on this machine —
/// one is mcpOAuth-only, the other has claudeAiOauth; we must enumerate all and pick
/// the right one rather than grabbing the first via `security find-generic-password -w`).

import Foundation
import Security

private let keychainService = "Claude Code-credentials"
private let credentialsPath = FileManager.default.homeDirectoryForCurrentUser
    .appendingPathComponent(".claude/.credentials.json")

// seconds-coercion guard — same constants as clud's token_resolver.py
private let msThreshold:    Int64 = 1_000_000_000_000  // values >= this are already ms
private let secondsFloor:   Int64 = 1_000_000_000       // plausible epoch-seconds lower bound

// MARK: - Public API

/// Result of a token resolution attempt.
public struct TokenResult: Equatable, Sendable {
    /// The access token, or nil on failure.
    public let token: String?
    /// Where the token came from: "env" | "file" | "keychain" | "none".
    public let source: String
    /// nil on success; "no_token" | "expired" on failure.
    public let reason: String?

    public init(token: String?, source: String, reason: String?) {
        self.token = token
        self.source = source
        self.reason = reason
    }
}

// MARK: - Pure / testable core

/// Extract (accessToken, expiresAt_ms) from a credentials JSON blob.
///
/// Shape: {"claudeAiOauth": {"accessToken": "...", "expiresAt": <epoch ms>}}.
/// Returns (nil, nil) for any malformed input — mirrors clud's _parse_oauth_blob.
public func parseOAuthBlob(_ text: String) -> (token: String?, expiresAtMs: Int64?) {
    guard let data = text.data(using: .utf8),
          let doc = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let oauth = doc["claudeAiOauth"] as? [String: Any]
    else { return (nil, nil) }

    let tokenVal = oauth["accessToken"] as? String
    let token: String? = (tokenVal?.isEmpty == false) ? tokenVal : nil

    var expiresAtMs: Int64?
    if let expiresNum = oauth["expiresAt"] as? NSNumber {
        let raw = expiresNum.int64Value
        // Guard: coerce seconds→ms only when value is a plausible epoch-seconds timestamp
        // in [1e9, 1e12). Smaller values (e.g. test sentinels) are taken as-is so a
        // clearly-past timestamp still reads as expired rather than inflated into the future.
        if raw >= secondsFloor && raw < msThreshold {
            expiresAtMs = raw * 1000
        } else {
            expiresAtMs = raw
        }
    }

    return (token, expiresAtMs)
}

/// Given a list of (account, blob) pairs from the Keychain, return the blob string
/// whose `parseOAuthBlob` yields a non-nil token.
///
/// WHY multi-item selection: on the target machine TWO generic-password items share
/// service "Claude Code-credentials" — account "unknown" (blob: mcpOAuth only) and
/// account "masongoetz" (blob: claudeAiOauth with accessToken). The `security
/// find-generic-password -w` approach (as used by clud) grabs whichever the OS picks
/// first, which is the wrong one here. We enumerate all items and prefer the account
/// matching the current username, then fall through by insertion order.
public func selectKeychainBlob(
    _ items: [(account: String, blob: String)],
    preferAccount: String
) -> String? {
    // Try the preferred account first.
    if let preferred = items.first(where: { $0.account == preferAccount }) {
        let (tok, _) = parseOAuthBlob(preferred.blob)
        if tok != nil { return preferred.blob }
    }
    // Fall through remaining items in order.
    for item in items where item.account != preferAccount {
        let (tok, _) = parseOAuthBlob(item.blob)
        if tok != nil { return item.blob }
    }
    return nil
}

/// Resolve a live OAuth token from injected sources — fully testable offline.
///
/// Priority: env → credentials file → keychain.
/// Re-call this on every fetch so token rotations by Claude Code are picked up
/// automatically (same discipline as clud's resolve_token).
public func resolveToken(
    env: [String: String],
    credentialsFileText: String?,
    keychainItems: [(account: String, blob: String)],
    preferAccount: String,
    nowMs: Int64
) -> TokenResult {
    // 1. Environment variable — highest priority, no expiry check (used by tests/CI).
    if let envTok = env["CLAUDE_CODE_OAUTH_TOKEN"], !envTok.isEmpty {
        return TokenResult(token: envTok, source: "env", reason: nil)
    }

    // 2. Credentials file: ~/.claude/.credentials.json
    var token: String?
    var expiresAtMs: Int64?
    var source = "none"

    if let fileText = credentialsFileText {
        let (t, e) = parseOAuthBlob(fileText)
        if t != nil {
            token = t
            expiresAtMs = e
            source = "file"
        }
    }

    // 3. Keychain: enumerate all items, prefer current username.
    if token == nil {
        if let blob = selectKeychainBlob(keychainItems, preferAccount: preferAccount) {
            let (t, e) = parseOAuthBlob(blob)
            if t != nil {
                token = t
                expiresAtMs = e
                source = "keychain"
            }
        }
    }

    guard let tok = token else {
        return TokenResult(token: nil, source: "none", reason: "no_token")
    }
    if let exp = expiresAtMs, exp <= nowMs {
        return TokenResult(token: nil, source: source, reason: "expired")
    }
    return TokenResult(token: tok, source: source, reason: nil)
}

// MARK: - Impure shell (used by the app)

/// Read all generic-password items for the Claude Code credentials service via
/// Security framework (NOT `security` CLI — GUI apps don't inherit shell PATH and
/// SecItemCopyMatching avoids the extra process + parsing).
///
/// First call may trigger a one-time macOS Keychain authorization prompt.
/// Any failure (denial, missing item, OS error) returns an empty array → caller
/// falls through to "no_token" degraded state.
func readKeychainItems(service: String = keychainService) -> [(account: String, blob: String)] {
    let query: [CFString: Any] = [
        kSecClass:            kSecClassGenericPassword,
        kSecAttrService:      service,
        kSecMatchLimit:       kSecMatchLimitAll,
        kSecReturnAttributes: true,
        kSecReturnData:       true,
    ]

    var result: CFTypeRef?
    let status = SecItemCopyMatching(query as CFDictionary, &result)
    guard status == errSecSuccess,
          let items = result as? [[CFString: Any]]
    else { return [] }

    return items.compactMap { item -> (String, String)? in
        guard let data = item[kSecValueData] as? Data,
              let blob = String(data: data, encoding: .utf8)
        else { return nil }
        let account = (item[kSecAttrAccount] as? String) ?? "unknown"
        return (account, blob)
    }
}

/// Live token resolution using the real environment, credentials file, and Keychain.
///
/// Injecting NSUserName() as preferAccount picks the correct keychain item on machines
/// where multiple items share the "Claude Code-credentials" service (verified on dev machine).
public func resolveTokenLive() -> TokenResult {
    let env = ProcessInfo.processInfo.environment
    let fileText = try? String(contentsOf: credentialsPath, encoding: .utf8)
    let items = readKeychainItems()
    let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
    return resolveToken(
        env: env,
        credentialsFileText: fileText,
        keychainItems: items,
        preferAccount: NSUserName(),
        nowMs: nowMs
    )
}
